File: Python/lists/lists.py
"""lists/lists.py

This file includes solutions to List Problems (1 - 28)

author(s): Konstantinos Railis <kons.railis@gmail.com>

"""
import itertools
import random


def last_element(list_a: list):
    """Problem 1: Find The Last Element of a List.

    Parameters
    ----------
    list_a : list
        The input list

    Returns
    -------
    element
        The last element of the list

    Raises
    ------
    TypeError
        If the given argument is not of `list` type
    ValueError
        If the input list does not contain elements

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')
    if len(list_a) < 1:
        raise ValueError('The input list does not contain any elements.')
    return list_a[-1]


def last_but_one(list_a: list):
    """Problem 2: Find the Last but One Element of a List.

    Parameters
    ----------
    list_a : list
        The input list

    Returns
    -------
    element
        The second to last element of the list

    Raises
    ------
    TypeError
        If the given argument is not of `list` type
    ValueError
        If the input list contains less than two elements

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')
    if len(list_a) < 2:
        raise ValueError('The input list contains less than two elements.')
    return list_a[-2]


def kth_element(list_a: list, k: int):
    """Problem 3: Find the K'th Element of a List

    Parameters
    ----------
    list_a : list
        The input list
    k : int
        The element to fetch

    Returns
    -------
    element
        The k'th element of the input list

    Raises
    ------
    TypeError
        If the given argument is not of `list` type
    ValueError
        If the input list contains less than two elements, or the given k is less than 1

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')
    if len(list_a) < k:
        raise ValueError(f'The input list contains less than [{k}] elements.')
    if k < 1:
        raise ValueError('The value of k cannot be less than 1.')
    return list_a[k - 1]


def length(list_a: list):
    """Problem 4: Find the number of Elements of a List.

    Parameters
    ----------
    list_a : list
        The input list

    Returns
    -------
    integer
        The length of the input list

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')
    return len(list_a)


def reverse(list_a: list):
    """Problem 5: Reverse a List.

    Parameters
    ----------
    list_a : list
        The input list

    Returns
    -------
    list_a : list
        The input list reversed

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')
    list_a.reverse()
    return list_a


def is_palindrome(list_a: list):
    """Problem 6: Find out whether a list is palindrome

    Parameters
    ----------
    list_a : list
        The input list

    Returns
    -------
    bool
        `True` if the list is palindrome, `False` otherwise

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')

    # Compare elements of the list symmetrically. If any two symmetric
    # elements differ then the list is not palindrome
    for i in range(0, int(len(list_a) / 2)):
        if list_a[i] != list_a[-1 - i]:
            return False
    return True


def flatten(list_a: list):
    """Problem 7: Flatten a Nested List Structure,

    Parameters
    ----------
    list_a : list
        The input list

    Returns
    -------
    flat : list
        The flattened input list

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')

    flat = []
    for x in list_a:
        if isinstance(x, list):
            # If element x is a list then the initial list should be
            # extended by the elements of list x
            flat.extend(flatten(x))
        else:
            # Otherwise a single element is appended
            flat.append(x)

    return flat


def compress(list_a: list):
    """Problem 8: Eliminate Consecutive Duplicates of List.

    Parameters
    ----------
    list_a : list
        The input list

    Returns
    -------
    list
        A list with consecutive duplicates eliminated.

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')

    if len(list_a) <= 1:
        return list_a

    compressed = [list_a[0]]
    for element in list_a:
        if compressed[-1] == element:
            continue
        compressed.append(element)

    return compressed


def pack(list_a: list):
    """Problem 9: Pack Consecutive Duplicates of a List.

    Parameters
    ----------
    list_a : list
        The input list

    Returns
    -------
    list of list
        A list of lists packed with consecutive duplicates

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')

    if len(list_a) <= 1:
        # In case of empty or one-element list return.
        return list_a

    packed, sublist = [], [list_a[0]]
    for element in list_a[1:]:
        if sublist[-1] != element:
            # If current element does not match the last of the sublist
            # extend packed list and reset sublist to a single-element
            packed.append(sublist)
            sublist = [element]
        else:
            # If another same element is found, append to the sublist
            sublist.append(element)
    packed.append(sublist)

    return packed


def encode(list_a: list):
    """Problem 10: Run Length Encoding of List.

    Parameters
    ----------
    list_a : list of list
        The input list

    Returns
    -------
    list of list
        An length-encoded list

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')

    # After packing the list using the previous function use the
    # sublist length and append the new pair to the output list
    encoded = []
    packed = pack(list_a)
    for x in packed:
        encoded.append([len(x), x[0]])

    return encoded


def encode_modified(list_a: list):
    """Problem 11: Modified Length Encoding of List.

    Parameters
    ----------
    list_a : list
        The input list

    Returns
    -------
    list
        A modified length-encoded list

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')

    encoded = []
    packed = pack(list_a)
    for x in packed:
        if len(x) > 1:
            encoded.append([len(x), x[0]])
        else:
            encoded.append(x[0])

    return encoded


def decode(encoded: list):
    """Problem 12: Decode a run-length encoded list.

    Parameters
    ----------
    encoded : list
        The encoded input list

    Returns
    -------
    list
        The decoded list

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(encoded, list):
        raise TypeError('The argument given is not of `list` type.')

    decoded = []
    for x in encoded:
        if isinstance(x, list):
            decoded.extend(x[0] * [x[1]])
        else:
            decoded.append(x)

    return decoded


def encode_direct(list_a: list):
    """Problem 13: Run-length encoding of a list (direct solution).

    Parameters
    ----------
    list_a : list
        The input list

    Returns
    -------
    list of list
        An length-encoded list

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')

    if len(list_a) <= 1:
        # In case of empty or one-element list return.
        return list_a

    encoded, current, count = [], list_a[0], 1
    for element in list_a[1:]:
        if current != element:
            # If current element does not match the recorded current
            # append the count to the list
            encoded.append(current if count == 1 else [count, current])
            current, count = element, 1
        else:
            # If another same element is found, increase counter
            count += 1
    encoded.append(current if count == 1 else [count, current])

    return encoded


def duplicate(list_a: list):
    """Problem 14: Duplicate the Elements of a list.

    Parameters
    ----------
    list_a : list
        The input list

    Returns
    -------
    list
        A list of duplicated elements in order

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')

    duplicated = []
    for x in list_a:
        duplicated.extend([x, x])

    return duplicated


def duplicate_times(list_a: list, times: int):
    """Problem 15: Duplicate the elements of a list a given number of times.

    Parameters
    ----------
    list_a : list
        The input list
    times : int
        The number of times to duplicate elements

    Returns
    -------
    list
        A list of n-times duplicated elements

    Raises
    ------
    TypeError
        If the given argument is not of `list` type
    ValueError
        If the given `times` value is smaller than or equal to zero

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')
    if times <= 0:
        raise ValueError('An invalid value was given for `times`.')

    if times == 1:
        return list_a

    duplicated = []
    for x in list_a:
        duplicated.extend(times * [x])
    return duplicated


def drop(list_a: list, period: int):
    """Problem 16: Drop every N'th element from a list

    Parameters
    ----------
    list_a : list
        The input list
    period : int
        The number of every N elements to drop

    Returns
    -------
    list
        A list without a number of dropped elements.

    Raises
    ------
    TypeError
        If the given argument is not of `list` type
    ValueError
        If the given `n` is greater than the list's length or smaller than 1

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')
    if period > len(list_a) or period < 1:
        raise ValueError('The value of `period` is not valid.')

    return [x for i, x in enumerate(list_a) if (i + 1) % period != 0]


def split(list_a: list, length_of_first: int):
    """Problem 17: Split a list in two parts; the length of the first list is given.

    Parameters
    ----------
    list_a : list
        The input list
    length_of_first : int
        The desired length of the first output list

    Returns
    -------
    (list, list)
        A tuple of lists

    Raises
    ------
    TypeError
        If the given argument is not of `list` type
    ValueError
        If the given `n` is greater than the list's length or smaller than 1

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')
    if length_of_first > len(list_a) or length_of_first < 0:
        raise ValueError('The value of `l` is not valid.')

    return list_a[:length_of_first], list_a[length_of_first:]


def slice_list(list_a: list, start: int, end: int):
    """Problem 18: Extract a slice from list.

    Parameters
    ----------
    list_a : list
        The input list
    start : int
        The start of the slice
    end : int
        The end of the slice

    Returns
    -------
    list
        A slice of the initial list

    Raises
    ------
    TypeError
        If the given argument is not of `list` type
    ValueError
        If the values of `start` and / or `end` are invalid

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')
    if start < 1 or len(list_a) < start:
        raise ValueError('The value of `start` argument is not valid')
    if end < start or len(list_a) < end:
        raise ValueError('The value of `end` argument is not valid')

    return list_a[(start - 1):end]


def rotate(list_a: list, places: int):
    """Problem 19: Rotate a List N places to the left.

    Parameters
    ----------
    list_a : list
        The input list
    places : int
        The number of places to rotate the list to the left

    Returns
    -------
    list
        A list rotated n places to the left.

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')

    places = places % len(list_a)
    rotated = list_a[places:]
    rotated.extend(list_a[:places])
    return rotated


def remove_kth_element(list_a: list, index: int):
    """Problem 20: Remove K'th element from list.

    Parameters
    ----------
    list_a : list
        The input list
    index : int
        The element number to remove from the list

    Returns
    -------
    list
        The input list without the removed element

    Raises
    ------
    TypeError
        If the given argument is not of `list` type
    ValueError
        If the given `k` is not valid.

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')
    if index < 0 or len(list_a) < index:
        raise ValueError('The argument `k` is not valid.')

    list_a.pop(index - 1)
    return list_a


def insert_at(list_a: list, position: int, item):
    """Problem 21: Insert element at a given position into a list.

    Parameters
    ----------
    list_a : list
        The input list
    position : int
        The position where the inserted element should be
    item
        The element to insert to the list

    Returns
    -------
    list
        A list with an inserted element at the proper position

    Raises
    ------
    TypeError
        If the given argument is not of `list` type
    ValueError
        If the given `k` is not valid.

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')
    if position < 0 or len(list_a) < position:
        raise ValueError('The value of `n` is not valid.')

    list_a.insert(position - 1, item)
    return list_a


def range_list(start: int, end: int):
    """Problem 22: Create a list containing all integers within a given range.

    Parameters
    ----------
    start : int
    end : int

    Returns
    -------
    list
        A list including all numbers from `start` to `end`

    Raises
    ------
    ValueError
        If the given `start` is smaller than `end`

    """
    if end < start:
        raise ValueError('The `end` argument is smaller than `start`.')

    return list(range(start, end + 1))


def rnd_select(list_a: list, n: int):
    """Problem 23: Extract a given number of randomly selected items from a list.

    Parameters
    ----------
    list_a : list
        The input list
    n : int
        The number of elements to randomly select

    Returns
    -------
    list
        A list of `n` randomly selected numbers out of the initial list

    Raises
    ------
    TypeError
        If the given argument is not of `list` type
    ValueError
        If the given `start` is smaller than `end`

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')
    if n < 0 or len(list_a) < n:
        raise ValueError('The value of the given `n` is invalid.')

    return random.sample(list_a, n)


def lotto(n: int, m: int):
    """Problem 24: Lotto; Draw N different random numbers from the set 1..M.

    Parameters
    ----------
    n : int
        The number of numbers to draw
    m : int
        The upper limit of the set to draw numbers from

    Returns
    -------
    list
        A list of randomly selected numbers from the set 1..m

    Raises
    ------
    ValueError
        If the value of `n` or `m` is invalid.

    """
    if n < 0 or m < 1:
        raise ValueError('The given values are invalid.')
    if m < n:
        raise ValueError('The value of numbers to draw cannot be greater than the range.')

    return rnd_select(range_list(1, m), n)


def rnd_permutation(list_a: list):
    """Problem 25: Generate Random Permutation of List.

    Parameters
    ----------
    list_a : list
        The input list

    Returns
    -------
    list
        A random permutation of the input list as a list

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')

    return rnd_select(list_a, len(list_a))


def combination(list_a: list, k: int):
    """Problem 26: Generate the combinations of K distinct objects chosen from N elements of list.

    Parameters
    ----------
    list_a : list
        The input list
    k : int
        The number of objects to use in each combination

    Returns
    -------
    list
        A list of combinations of k elements out of the input list

    Raises
    ------
    TypeError
        If the given argument is not of `list` type
    ValueError
        If the given `k` is not valid.

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')
    if k < 1 or len(list_a) < k:
        raise ValueError('The given value of `k` is not valid.')

    return list(itertools.combinations(list_a, k))


def disjoint_subsets(list_a: list, groups):
    """Problem 27: Group elements of set into disjoint subsets.

    Parameters
    ----------
    list_a : list
        The input list
    groups

    Returns
    -------

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')

    set_a = set(list_a)
    # Find all combinations with the given plurality of the first subset
    first_set = itertools.combinations(set_a, groups[0])
    disjoint_list = []
    for x in first_set:
        # For every subset of the possible combinations of the first
        # group, remove its members and find all combinations with
        # given plurality of the second subset
        temp_subset = set_a - set(x)
        second_set = itertools.combinations(temp_subset, groups[1])
        list_x = list(x)
        for y in second_set:
            subset = [list_x, list(y), list(temp_subset - set(y))]
            disjoint_list.append(subset)
    return disjoint_list


def length_sort(list_a: list):
    """Problem 28: (a) Sort lists according to length of sublists

    Parameters
    ----------
    list_a : list
        The input list

    Returns
    -------
    list of list
        A list sorted according to length of sublists

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')

    return sorted(list_a, key=len)


def length_frequency_sort(list_a: list):
    """Problem 28: (b) Sort lists according to length frequency of sublists

    Parameters
    ----------
    list_a : list
        The input list

    Returns
    -------
    list of list
        A list of lists sorted according to the length frequency of the sublists

    Raises
    ------
    TypeError
        If the given argument is not of `list` type

    """
    if not isinstance(list_a, list):
        raise TypeError('The argument given is not of `list` type.')

    sorted_list = []
    list_length_dict = dict()
    for list_length in list(map(len, list_a)):
        list_length_dict[list_length] = [y for y in list_a if len(y) == list_length]
    length_sorted = sorted(list_length_dict.values(), key=len)
    for sublist in length_sorted:
        sorted_list.extend(sublist)

    return sorted_list

File: Golang/pkg/lists/lists.go
// Package lists includes solutions to List Problems (1-28).
package lists

import (
	"errors"
	"math/rand"
	"sort"
	"time"
)

type Slice []string
type ISlice []interface{}
type LengthEncodedPair struct {
	length int
	str    string
}

type ByLength []Slice

func (a ByLength) Len() int           { return len(a) }
func (a ByLength) Less(i, j int) bool { return len(a[i]) < len(a[j]) }
func (a ByLength) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }

type ByLength2 [][]Slice

func (a ByLength2) Len() int           { return len(a) }
func (a ByLength2) Less(i, j int) bool { return len(a[i]) < len(a[j]) }
func (a ByLength2) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }

// LastElement find The Last Element of a Slice (P01).
func LastElement(s Slice) (string, error) {
	length := len(s)
	if length < 1 {
		return "", errors.New("The given list is empty. Try again with a non-empty list.")
	}
	return s[length-1], nil
}

// LastButOneElement finds the Last but One Element of a Slice (P02).
func LastButOneElement(s Slice) (string, error) {
	length := len(s)
	if length < 2 {
		return "", errors.New("The given list has fewer elements than needed.")
	}
	return s[length-2], nil
}

// KthElement finds the k'th Element of a Slice (P03).
func KthElement(s Slice, k int) (string, error) {
	length := len(s)
	if length < k {
		return "", errors.New("The given list's length is smaller than the given k.")
	}
	return s[k-1], nil
}

// Length finds the lenght of a Slice (P04).
func Length(s Slice) int {
	return len(s)
}

// Reverse reverses a given Slice (P05).
func Reverse(s Slice) Slice {
	length := len(s)
	for i := 0; i < length/2; i++ {
		s[i], s[length-i-1] = s[length-i-1], s[i]
	}
	return s
}

// IsPalindrome returns true if a given Slice is palindrome (P06).
func IsPalindrome(s Slice) bool {
	length := len(s)
	for i := 0; i < length/2; i++ {
		if s[i] != s[length-i-1] {
			return false
		}
	}
	return true
}

// Flatten flattens a nested ISlice structure (P07).
func Flatten(s ISlice) Slice {
	var result Slice
	for _, elem := range s {
		switch x := elem.(type) {
		case string:
			result = append(result, x)
		case ISlice:
			result = append(result, Flatten(x)...)
		}
	}
	return result
}

// Compress removes consecutive duplicates of a Slice (P08).
func Compress(s Slice) Slice {
	if len(s) < 1 {
		return s
	}
	result := Slice{s[0]}
	for _, elem := range s {
		if result[len(result)-1] == elem {
			continue
		}
		result = append(result, elem)
	}
	return result
}

// Pack packs consecutive elements and returns a Slice of Slice (P09).
func Pack(s Slice) []Slice {
	var result []Slice
	if len(s) == 0 {
		return []Slice{}
	}
	sublist := Slice{s[0]}
	for _, elem := range s[1:] {
		if sublist[len(sublist)-1] != elem {
			result = append(result, sublist)
			sublist = append(Slice{}, elem)
		} else {
			sublist = append(sublist, elem)
		}
	}
	result = append(result, sublist)
	return result
}

// Encode returns a length-encoded Slice (P10).
func Encode(s Slice) ISlice {
	result := ISlice{}
	packedSlice := Pack(s)
	for _, elem := range packedSlice {
		result = append(result, LengthEncodedPair{len(elem), elem[0]})
	}
	return result
}

// EncodeModified returns a modified length encoding of a Slice (P11).
func EncodeModified(s Slice) ISlice {
	result := ISlice{}
	packedSlice := Pack(s)
	for _, elem := range packedSlice {
		if len(elem) == 1 {
			result = append(result, elem[0])
		} else {
			result = append(result, LengthEncodedPair{len(elem), elem[0]})
		}
	}
	return result
}

// Decode decodes a length-encoded Slice (P12).
func Decode(encoded ISlice) Slice {
	decoded := Slice{}
	for _, elem := range encoded {
		switch x := elem.(type) {
		case string:
			decoded = append(decoded, x)
		case LengthEncodedPair:
			for i := 0; i < x.length; i++ {
				decoded = append(decoded, x.str)
			}
		}
	}
	return decoded
}

// EncodeDirect offers an alternative solution to Encode (P13).
func EncodeDirect(s Slice) ISlice {
	if len(s) == 0 {
		return ISlice{}
	} else if len(s) == 1 {
		return ISlice{s[0]}
	}

	encoded, current, count := ISlice{}, s[0], 1
	for _, elem := range s[1:] {
		if current != elem {
			if count == 1 {
				encoded = append(encoded, current)
			} else {
				encoded = append(encoded, LengthEncodedPair{count, current})
			}
			current, count = elem, 1
		} else {
			count += 1
		}
	}
	if count == 1 {
		encoded = append(encoded, current)
	} else {
		encoded = append(encoded, LengthEncodedPair{count, current})
	}
	return encoded
}

// Duplicate doubles the elements of the list (P14).
func Duplicate(s Slice) Slice {
	duplicated := Slice{}
	for _, elem := range s {
		duplicated = append(duplicated, elem, elem)
	}
	return duplicated
}

// DuplicateTimes duplicates the elements of the list a given number of times (P15).
func DuplicateTimes(s Slice, times int) Slice {
	if times == 1 {
		return s
	}

	duplicated := Slice{}
	for _, elem := range s {
		for i := 0; i < times; i++ {
			duplicated = append(duplicated, elem)
		}
	}
	return duplicated
}

// Drop removes every N'th element of a list (P16).
func Drop(s Slice, period int) (Slice, error) {
	if period > len(s) || period < 1 {
		return nil, errors.New("The value of `period` is not valid.")
	}
	if period == 1 {
		return Slice{}, nil
	}
	result := Slice{}
	for i := 0; i < len(s); i++ {
		if (i+1)%period != 0 {
			result = append(result, s[i])
		}
	}
	return result, nil
}

// Split splits a list in two parts according to the length of the first (P17).
func Split(s Slice, lengthOfFirst int) (Slice, Slice, error) {
	if lengthOfFirst > len(s) || lengthOfFirst < 0 {
		return nil, nil, errors.New("The length of the first list is not valid.")
	}
	return s[:lengthOfFirst], s[lengthOfFirst:], nil
}

// SliceList returns a Slice from a list (P18).
func SliceList(s Slice, start int, end int) (Slice, error) {
	if start < 1 || len(s) < start {
		return nil, errors.New("The value of `start` is not valid.")
	}
	if end < start || len(s) < end {
		return nil, errors.New("The value of `end` is not valid.")
	}
	return s[start-1 : end], nil
}

// Rotate rotates a list N places to the left (P19).
func Rotate(s Slice, places int) Slice {
	places = places % len(s)
	if places < 0 {
		places = len(s) + places
	}
	rotated := s[places:]
	rotated = append(rotated, s[:places]...)
	return rotated
}

// RemoveAt removes the K'th element of a list (P20).
func RemoveAt(s Slice, index int) (Slice, error) {
	if index < 0 || len(s) < index {
		return nil, errors.New("The `index` argument is not valid.")
	}
	return append(s[:index-1], s[index:]...), nil
}

// InsertAt inserts an element at a given position (P21).
func InsertAt(s Slice, position int, element string) (Slice, error) {
	if position < 0 || len(s)+1 < position {
		return nil, errors.New("The `position` argument is not valid.")
	}
	if position == len(s)+1 {
		return append(s, element), nil
	}
	s = append(s[:position], s[position-1:]...)
	s[position-1] = element
	return s, nil
}

// Range returns a list of integers from start to end (P22).
func Range(start int, end int) ([]int, error) {
	if end < start {
		return nil, errors.New("The given arguments are invalid.")
	}
	rangeList := []int{}
	for i := start; i <= end; i++ {
		rangeList = append(rangeList, i)
	}
	return rangeList, nil
}

// RandomSelect picks N random elements from a List (P23).
func RandomSelect(s []int, samples int) ([]int, error) {
	if samples < 0 || len(s) < samples {
		return nil, errors.New("The requested number of samples is invalid.")
	}
	rand.Seed(time.Now().UnixNano())
	randomlySelected := []int{}
	for i := 0; i < samples; i++ {
		randomlySelected = append(randomlySelected, s[rand.Intn(len(s))])
	}
	return randomlySelected, nil
}

// Lotto randomly picks N integers from Range 1 to M (P24).
func Lotto(n int, m int) ([]int, error) {
	if n < 0 || m < 1 {
		return nil, errors.New("The given values are invalid.")
	}
	if m < n {
		return nil, errors.New("The value of numbers to draw cannot be greater than the range.")
	}
	rangeList, _ := Range(1, m)
	return RandomSelect(rangeList, n)
}

// RandomPermutation generates a random permutation of a list (P25).
func RandomPermutation(s []int) ([]int, error) {
	return RandomSelect(s, len(s))
}

// Combination returns the combinations of K distinct elements of a list (P26).
func Combination(s Slice, k int) <-chan Slice {
	channel := make(chan Slice)
	go func() {
		defer close(channel)
		if k <= 0 {
			channel <- Slice{}
			return
		}
		thisone := Slice{}
		for i := 0; i < len(s); i++ {
			thisone = Slice{s[i]}
			for another := range Combination(s[i+1:], k-1) {
				channel <- append(thisone, another...)
			}
		}
	}()
	return channel
}

// LengthSort sorts a List according to the length of its sublists (P28a)
func LengthSort(s []Slice) []Slice {
	sort.Sort(ByLength(s))
	return s
}

// LengthFrequencySort sorts a List accoring to sublist length frequency (P28b)
func LengthFrequencySort(s []Slice) []Slice {
	// Group sublists according to length in a map
	listLenMap := make(map[int][]Slice)
	for _, elem := range s {
		if value, ok := listLenMap[len(elem)]; ok {
			listLenMap[len(elem)] = append(value, elem)
		} else {
			listLenMap[len(elem)] = []Slice{elem}
		}
	}
	// Keep only grouped sublists
	groupedSublists := [][]Slice{}
	for _, sublists := range listLenMap {
		groupedSublists = append(groupedSublists, sublists)
	}
	sort.Sort(ByLength2(groupedSublists))
	// Form output slice
	sorted := []Slice{}
	for _, sublist := range groupedSublists {
		sorted = append(sorted, sublist...)
	}
	return sorted
}
